exasol/analytics/audit/audit_query_handler.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from enum import Enum
+from typing import (
+    Callable,
+    Generic,
+    Iterator,
+    Optional,
+    TypeAlias,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from pydantic import BaseModel
+
+from exasol.analytics.audit.audit import AuditTable
+from exasol.analytics.query_handler.context.scope import ScopeQueryHandlerContext
+from exasol.analytics.query_handler.graph.stage.sql.sql_stage_graph import SQLStageGraph
+from exasol.analytics.query_handler.query.interface import Query
+from exasol.analytics.query_handler.query.result.interface import QueryResult
+from exasol.analytics.query_handler.query.select import (
+    AuditQuery,
+    SelectQueryWithColumnDefinition,
+)
+from exasol.analytics.query_handler.query_handler import QueryHandler
+from exasol.analytics.query_handler.result import (
+    Continue,
+    Finish,
+)
+
+
+class Phase(Enum):
+    MAIN = 2
+    RUN_FINAL_AUDIT_QUERY = 3
+    FINISHED = 4
+
+
+ResultType = TypeVar("ResultType")
+ParameterType = TypeVar("ParameterType")
+ContinueOrFinish: TypeAlias = Union[Continue, Finish[ResultType]]
+
+
+class IllegalMethodCallError(RuntimeError):
+    """
+    The application framework logic called a method that was not intended
+    to be called, e.g. ``IdentityQueryHandler.handle_query_result()``.
+    """
+
+
+@dataclass
+class Child(Generic[ResultType]):
+    context: ScopeQueryHandlerContext
+    query_handler: QueryHandler
+    result: ResultType | None
+
+
+class AuditQueryHandler(QueryHandler[ParameterType, ResultType]):
+    """
+    Use the specified factory (e.g. :class:`OrchestratorQueryHandler`) to
+    instantiate a query handler and augment the queries returned by it for
+    creating Audit Log entries.
+    """
+
+    def __init__(
+        self,
+        parameter: ParameterType,
```

Comment:
class AuditQueryHandlerParameter[ParameterType]:

child_parameter:ParameterType
db_schema
table_prefix

Reply:
Theoretical the factory could contain the parameter which would make the parameter object easier

Reply:
Alternative, we don't have AuditQueryHandlerParameter, but we have functions as parameters, that extract db_schema and table_prefix from whatever Parameter we got

Reply:
See last push